selectrum.el

```diff
@@ -1258,33 +1328,110 @@ list). A null or non-positive ARG inserts the candidate corresponding to
+  (interactive)
+  (setq selectrum-history-use-input
+        (not selectrum-history-use-input))
+  (minibuffer-message "Toggled to %s history"
+                      (if selectrum-history-use-input
+                          "input"
+                        "selection"))
+  (when (> minibuffer-history-position 0)
+    (delete-minibuffer-contents)
+    (insert
+     (nth (1- minibuffer-history-position)
+          (selectrum--minibuffer-history-value)))))
+
+(defun selectrum-previous-history-element (n)
+  "Forward to `previous-history-element'.
+With argument N, it uses the Nth previous element."
+  (interactive "p")
+  (let ((minibuffer-history-variable
+         (selectrum--get-current-history-var)))
+    (previous-history-element n)))
+
+(defun selectrum-next-history-element (n)
+  "Forward to `next-history-element'.
+With argument N, it uses the Nth following element."
+  (interactive "p")
+  (let ((minibuffer-history-variable
+         (selectrum--get-current-history-var)))
+    (next-history-element n)))
+
 (defun selectrum-select-from-history ()
-  "Select a candidate from the minibuffer history.
-If Selectrum isn't active, insert this candidate into the
-minibuffer."
+  "Submit or insert candidate from minibuffer history.
+If Selectrum isn't active, don't exit after submission."
   (interactive)
-  (let ((selectrum-should-sort-p nil)
-        (enable-recursive-minibuffers t)
-        (history (symbol-value minibuffer-history-variable)))
+  (let ((history (let ((selectrum-history-use-input nil))
+                   (selectrum--minibuffer-history-value))))
     (when (eq history t)
       (user-error "No history is recorded for this command"))
     (let ((result
-           (let ((selectrum-candidate-inserted-hook nil)
-                 (selectrum-candidate-selected-hook nil))
-             (selectrum-read "History: " history :history t))))
-      (if (and selectrum--match-required-p
-               (not (member result selectrum--refined-candidates)))
-          (user-error "That history element is not one of the candidates")
-        (if selectrum-active-p
-            (selectrum--exit-with result)
-          (insert result))))))
+           (let* ((input/history
+                   (let ((selectrum-history-use-input t))
+                     (selectrum--minibuffer-history-value)))
+                  (history (cl-mapcar
+                            (lambda (item input)
+                              (if (or (equal item input)
+                                      (string-empty-p input))
+                                  (propertize item 'input item)
+                                ;; Allow to match agains both.
+                                (propertize
+                                 (concat item
+                                         (propertize " -- " 'face 'bold)
+                                         input)
+                                 'selectrum-candidate-full item
+                                 'input input)))
```

Comment:
I tried to improve it like you suggested. I could have also used `selectrum-candidate-display-suffix` but I thought in this case it would be nice to be able to match against the input history, too.

Reply:
Fair enough, I could go either way on that.